Basic draft for eulerian paths/circuits

YasirKusay opened the pull request:
I started to implement eulerian cycles/paths. I made:

- .c file in src with a few functions, will refine later on
- .h in include, containing only one function for now
- a .c file in examples/tests
- a .at file inside tests

I am not sure as how I can include my files into the compiling and tests steps, can you please help me?
After that, I will refine the code that I have.

szhorvat commented:
> I am not sure as how I can include my files into the compiling and tests steps, can you please help me?

You can add the .c and .h files to `src/Makefile.am`, to the `SOURCES` and `HEADERS_PUBLIC` sections, respectively. Before you can do this, any `main` must be removed.

I know that you are still working on this, but when it is done there must be no printing in the code.

szhorvat commented:
A few more comments:

 - Please stick to the C89 standard (notice `//` comments and variable definitions not at the start of the block).
 - Follow igraph's function naming conventions (`igraph_is_eulerian()`)
 - Any igraph function that can fail in any way must return an error code (not a result). Results are returned through arguments passed by reference.
 - Arguments passed by reference that won't be modified must be `const` (`const igraph_t *graph`)
 - `igraph_t copy = *graph` does *not* make a copy. See `igraph_copy()`.

YasirKusay commented:
Hello, I apologise for my inactivity and I will get back to working on this now.

YasirKusay commented:
Hello, I am very close to finishing but when calling make, there are no warnings/errors but there is a "linker command failed with exit code 1 issue", could you help me with this issue?

szhorvat commented:
Not without more details. We need to see what you have.

YasirKusay commented:
<img width="1675" alt="Screen Shot 2020-03-22 at 9 45 22 pm" src="https://user-images.githubusercontent.com/59812220/77247554-799d7080-6c86-11ea-8dfb-9403aaae82f4.png">

szhorvat commented:
The linking error relates to address sanitizer. Have you tried compiling without that?

It's not really possible to say more without knowing what you are doing exactly (each step described precisely).

YasirKusay commented:
I am following the compilation steps on: https://igraph.org/c/#docs
./configure
make
make check
make install

./configure worked fine, but when calling make (the next step), it presented me with the error above

szhorvat commented:
AddressSanitizer is not enabled by default. Are you certain that this is precisely what you did and nothing else? I do not see how ASAN would get in there otherwise.

Did you do a `make distclean` before changing any ./configure arguments? Did you change any compiler flags (CFLAGS variable or similar)?

Please try to do it again with a clean source tree.

YasirKusay commented:
Hello,  ive compiled my file and have begun to perform testing on it. One of the issues that I am getting now is that in the testfile for which I am calling my newly written function, it says implicit declaration, even though ive declared the function prior, (in the .h file). Is there another place where I may need to declare the function in?

YasirKusay commented:
I apologise for asking a pretty trivial issue here, however, when calling one of the functions that I wrote into my test file, it says implicit declaration of that function.
<img width="1280" alt="Screen Shot 2020-03-24 at 12 59 14 pm" src="https://user-images.githubusercontent.com/59812220/77380487-474b5a80-6dcf-11ea-8994-65b2fc346663.png">
Ive made sure to have all the necessary includes in my files and also to have declared my new functions and function file in all necessary locations.
Are there any locations that I may be missing? What would you recommend I do?

szhorvat commented:
> Ive made sure to have all the necessary includes in my files and also to have declared my new functions and function file in all necessary locations.

The error message says that this is not the case. The declaration is not included into this test file.

It is not possible to say more than this without seeing what you have. If you are stuck, I suggest you push the changes.

YasirKusay commented:
Ive push all my changes to you, hopefully you can see them now

szhorvat commented:
The prototype is not included in your test file: it is only present in `igraph_eulerian.h`, but that file is not included in `igraph.h` (it should be). You only have `#include <igraph.h>` in your test file. So the contents of `igraph_eulerian.h` are not visible from your test file.

`#include` literally includes the entire contents of the included file. The prototype must be in some way included in the file where the function is going to be used.

Are you familiar with how includes and prototypes work?

----

Let us know when you are ready to start receiving feedback.

YasirKusay commented:
Thank you for your reply, I am aware of how includes and prototypes work. I feel silly now as I forgot to include it into igraph.h.

szhorvat commented:
One more thing that's worth mentioning now:

 - When calling any function that returns an error code, it must be wrapped in `IGRAPH_CHECK`
 - Any function that has `IGRAPH_CHECK` in it *must* return an error code (because that's what `IGRAPH_CHECK` does: it `return`s with an error code when something goes wrong)

YasirKusay commented:
Hello, I am pretty close to completion. The is_eulerian function works as expected and the function to find eulerian paths works partially, for undirected graphs. For directed graphs, testing them separately worked fine but in the igraph_eulerian file that I have committed, I am getting an assertion failure (where no < 100) in the igraph_finally_real function and I traced that issue to the delete_edges function. I am going to stop working on it now and will resume working on it tomorrow after work. I believe that I can finish by the end of tomorrow.

Furthermore, I would like to add that I am storing paths as igraph_vector_t momentarily and I will change it to a more appropriate type at the end.

szhorvat commented:
> I am getting an assertion failure (where no < 100) in the igraph_finally_real function

This is usually an indication that the number of `IGRAPH_FINALLY`s does not match the number of `IGRAPH_FINALLY_CLEAN`s.

YasirKusay commented:
Could you help me understand that issue in the context of my code? By looking closer, I noticed that the program breaks with the above invocation (assertion failure where no < 100) on line 317 of igraph_eulerian.c where I am calling igraph_delete_edges(graph, es); I call the same function for the non directed graph case and the worked fine but in the case of this, it presents me issues. 
<img width="1110" alt="Screen Shot 2020-03-25 at 7 42 23 pm" src="https://user-images.githubusercontent.com/59812220/77517755-c2e10080-6ed0-11ea-9b62-f8065db187e2.png">

szhorvat commented:
If you are not sure what a certain function or macro does, please ask instead of copying usage patterns from elsewhere. Doing that blindly will lead to errors that will be hard to fix later.

IGRAPH_FINALLY and related macros are documented here:

https://igraph.org/c/doc/igraph-Error.html#igraph-deallocatingmemory

The purpose of IGRAPH_FINALLY is to register a data structure for deallocation in case of error or interruption. If the function succeeded, the registred deallocation commands must be cleaned up with IGRAPH_FINALLY_CLEAN. Here you use IGRAPH_FINALLY, but there is no matching IGRAPH_FINALLY_CLEAN at the end of the function. Since you registered two things, you need IGRAPH_FINALLY_CLEAN(2). Be very careful with this as with complicated algorithms it can be hard to keep track of how many things have been registered. It is critical that precisely the same number are cleaned.

szhorvat commented:
Yes, having to use these macros is pretty annoying, especially if you are familiar with exceptions in C++. But in C, we don't really have any better choices. Sometimes I really wish igraph were written in C++ ...

szhorvat commented:
Should this have been a `1`?

https://github.com/igraph/igraph/pull/1346/commits/b9ccea6bd11949f2922f24d41169302b95ed909f#diff-272142b27a0fb1dab2fdde056555e68cR275

YasirKusay commented:
(Thanks for pointing the above out).

It appears to be working fine now, you can go ahead and give me feedback. 
There is something I am interested in though, I really like the igraph_vector_t type and find that it is very flexible to work with (I store all paths inside it). However, it would make more sense to store the result as an int (it would make no sense to denote a vertice as a float). Are there any similar functions to igraph_vector_t that I can use?

szhorvat commented on examples/tests/igraph_eulerian_paths.c:
```diff
@@ -0,0 +1,79 @@
+
+int main() {
+
+    igraph_t graph, graph2, graph3, graph4, graph5, graph6, graph7, graph8, graph9;
```
You can use a single variable `graph` which you destroy as soon as you are done with one graph, and create a new graph in its place. It is not necessary to have 9 different ones.

iosonofabio replied:
This got solved in the meanwhile.

szhorvat commented on include/igraph_eulerian.h:
```diff
@@ -0,0 +1,51 @@
+#include "igraph_datatype.h"
+#include "igraph_iterators.h"
+#include "igraph_attributes.h"
+#include "igraph_sparsemat.h"
```
Most of these `#include`s are not needed. Use only those which are required by the header itself. If any are required only by the `.c` file, add them only to the `.c` file, not the header.

szhorvat commented on include/igraph_eulerian.h:
```diff
@@ -0,0 +1,51 @@
+__BEGIN_DECLS
+
+DECLDIR int igraph_is_eulerian(igraph_t *graph);
+DECLDIR int is_eulerian_undirected(igraph_t *graph);
```
If a function is meant to be used only internally by the implementation, it should not be present the public header.

If a function is meant to be part of the public API, it should follow the naming convention of other igraph functions.

iosonofabio replied:
done by now

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,415 @@
+0 -> if graph is not Eulerian
+1 -> if graph has an euler circuit
+2 -> if graph has an euler path */
+int is_eulerian_undirected(igraph_t *graph) {
```
Any function with `IGRAPH_CHECK` inside _must_ return an error code.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,415 @@
+
+    odd = 0;
+
+    igraph_is_connected(graph, &res, IGRAPH_WEAK);
```
IGRAPH_CHECK is missing from here (and from many other places)

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,415 @@
+    int i;
+
+    if (igraph_is_eulerian(graph) == 0) {
+        IGRAPH_WARNING("Euler cycle not possible");
```
When returning an error code, use `IGRAPH_ERROR` (not `IGRAPH_WARNING`)

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,415 @@
+    igraph_is_connected(graph, &res_strong, IGRAPH_STRONG);
+    igraph_is_connected(graph, &res_weak, IGRAPH_WEAK);
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong == 1)) {
```
If `b` is a boolean, please use `if (b)` and not `if (b == 1)`.

This is more readable, and generally "true" is represented with *a non-zero value* (not with 1).

szhorvat commented:
It looks like the functions return vertex IDs. An Eulerian path (circuit) consists of edges, not vertices. A list of edge IDs should be returned, not vertex IDs.

szhorvat commented:
I added a few comments (not a review, as there's still a lot to be done). Most comments apply to many parts of the code, not just where they were added.

As for storing vertex or edge IDs as `double` instead of integers, yes, this is not good, and it is a leftover from earlier days of igraph. We plan to fix this throughout the code base for version 1.0. For now, please continue returning an `igraph_vector_t` because any function that expects vertex/edge IDs as input takes this type. In other situations (e.g. internal use), if you need an integer vector, there's `igraph_vector_int_t` (which you already used).

YasirKusay commented:
Thank you for your extensive and speedy feedback. I am happy to change the implementation such that edge id's are stored, rather than vertex id's. Rather than returning the result as just edge id's, would you like to have an extra feature (perhaps an extra parameter in the function), giving the user the option to return the result as vertex id's or edge id's?

szhorvat commented:
> Rather than returning the result as just edge id's, would you like to have an extra feature (perhaps an extra parameter in the function), giving the user the option to return the result as vertex id's or edge id's?

It wouldn't hurt I guess ... but my opinion is that it is better to stick to returning edges only, to avoid user confusion.

Returning vertices is a very minor convenience; one can easily compute this with good performance from the edges. If we wanted to provide such a convenience, it would better be implemented as a separate function that transforms an edge-walk to a vertex-walk.

----

On a related note, did you verify the algorithm for multigraphs and graph with self-loops? We should aim for an implementation that supports both.

YasirKusay commented:
I guess it would make sense to store edge id's rather than vertex id's. I will modify it as such.
As per the second question, I have not but plan on expanding the tests.

szhorvat commented:
Also consider what happens when calling the function with the null graph (no vertices) or with a non-connected graph.

szhorvat commented:
Can you give a summary of the proposed public API (just post the prototypes here in the comments and explain the meaning of any arguments if they're not obvious)?

YasirKusay commented:
igraph_is_eulerian(igraph_t *graph) -> pretty much takes any graph and determines if it is eulerian or not. The function automatically determines if it is directed or not.
igraph_eulerian_paths(igraph_t *graph, igraph_vector_t *res) -> takes any graph and finds the eulerian path/cycle and stores that in res. Similar to the above, it automatically determines if the graph is directed or not.

szhorvat commented:
Does it find a *path* or a *cycle*?

YasirKusay commented:
Just a single path, it can also find cycles

szhorvat commented:
@YasirKusay  This is not what I was asking for. You posted what you have now, without addressing any of the issues I brought up 4 days ago. This is not going to work.

Before you go on to fix anything, it is a good idea to do a little planning.

I suggest you plan _the final API_, which hopefully will need no more adjustments,  and show us what you propose. Make sure you address all the issues.  Describe anything that is not obvious (i.e. write a small version of the documentation, what is each argument, what is the return value).

 - You didn't address the return value issue with `igraph_is_eulerian`. Also, document the meaning of the result—what does "Eulerian" mean? Address a naïve user, new to the concept. 
 - Why `paths` and not `path` if returning one?
 - How does a naïve user know if they are going to get a path or a cycle?
 - What is being returned through the `*res` argument?

iosonofabio commented:
@YasirKusay why don't we skype tomorrow afternoon and go through these together? It might be easier that way

YasirKusay commented:
Hello, I just spoke with Fabio and here is what we decided that I should do:
Here is the proposed public API:
igraph_is_eulerian(igraph_t *graph) -> this takes a pointer to the graph determines if it has a path, cycle, or neither, and the function will return 2, 1 or 0 respectively. It can do this for both directed and undirected graphs. An eulerian path is a trail that visits exactly each edge once, and a eulerian cycle is also a trail that visits exactly each edge once, but also starts and ends on the same vertex.

I understand that the user may be confused by the use of "paths", as it implies that my function is going to return multiple paths when it will only return 1. Therefore:
igraph_eulerian_path(igraph_t *graph, igraph_vector_t *res) -> this takes a pointer to a graph to which it will find an eulerian path for, and store the edge id trail in res. It will return an error if there are no paths in the graph.

Furthermore, although the way I have written the function "igraph_eulerian_path" to be able to find paths and cycles, I understand that a user may be confused by this and not know if they are getting a path or a cycle, and therefore, I will write an extra function:
igraph_eulerian_cycle(igraph_t *graph, igraph_vector_t *res), taking a pointer to a graph to which it will find a eulerian cycle for, and store the edge id trail in res. It will return an error if there are no cycles in the graph.

I will modify the implementation of igraph_eulerian_path/igraph_eulerian_cycle such that it will store edge id's rather than vertex id's.

In the future, if these functions get merged, I can try to write extra functions, e.g. igraph_eulerian_paths, that will return more than one eulerian path, and igraph_eulerian cycles, that will return more than one eulerian cycles.

Finally, with respect the the failed checks, it appears that this is because of a small issue to do with linux that I have not had in mac, where I wrote my implementation. I will make a commit today that will fix the issue.

I hope this has been of help, if you have any extra questions, please let me know.

szhorvat commented:
Your current implementation of `igraph_is_eulerian` uses functions that can fail. Therefore, `igraph_is_eulerian` itself can fail. As I said above in several messages, this means that it *must* return an error code. In fact, you currently use `IGRAPH_CHECK` in its implementation, and `IGRAPH_CHECK` effectively includes `return error_code`.

Use the following prototoype:

    int igraph_is_eurlerian(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle)

The return value is an error code. The existence of paths/cycles is returned through two output arguments.

YasirKusay commented:
I understand what you mean now, and I will change it such that it follows your prototype.

iosonofabio commented:
good idea @szhorvat that signature is even better

szhorvat commented on include/igraph_eulerian.h:
```diff
@@ -36,8 +36,8 @@
-DECLDIR int igraph_is_eulerian(igraph_t *graph);
-DECLDIR int igraph_eulerian_paths(igraph_t *graph, igraph_vector_t *res);
+DECLDIR igraph_integer_t igraph_is_eulerian(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle);
+DECLDIR igraph_integer_t igraph_eulerian_paths(igraph_t *graph, igraph_vector_t *res);
```
The return type should remain `int`. Error codes are always `int` in igraph. (You can take a look at some existing igraph functions for reference.)

szhorvat commented on examples/tests/igraph_eulerian_paths.c:
```diff
@@ -0,0 +1,79 @@
+
+    igraph_small(&graph, 0, IGRAPH_UNDIRECTED, 0,1 , 1,2, -1);
+    igraph_eulerian_paths(&graph, &res);
+    print_vector(&res, stdout);
```
Use `print_vector_round` when the elements of the vector are integer valued. You can look in `test_utilities.inc` to see what is available.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -415,11 +415,78 @@ igraph_integer_t igraph_eulerian_path_directed(igraph_t *graph, igraph_vector_t
+            vec1 = VECTOR(vector_res)[i];
+            vec2 = VECTOR(vector_res)[i+1];
+
+            igraph_get_eid(graph, &curr_edge, (int) vec1, (int) vec2, 1, error);
```
Repeated call to `igraph_get_eid` are going to be extremely slow. They must do a full search for the edge. Can you avoid using this function?

Also, you are casting to `int` when the expected argument is `igraph_integer_t`. Do not assume that these are the same. Use the same type that is required for the argument.

YasirKusay replied:
Thank you for your reply. To be honest, I do not really like this implementation as well (especially when operating on especially big graphs) but at the moment, I do not really have any other way of doing this. I really wanted to do this process (getting the edge id's) as the eulerian walk process is happening, but I soon realised that deleting an edge will also change all the edge id's, so I decided not to do this. If I have a better idea, I will definitely change this.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+    odd = 0;
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res, IGRAPH_WEAK));
+    if (res == 0) {
```
`res` is a Boolean. Use `if (!res)`, not `if (res == 0)`.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_ALL));
```
Is there any reason you are building a full inclist just to check the degree? We have `igraph_degree`.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+0 -> if graph is not Eulerian
+1 -> if graph has an euler circuit
+2 -> if graph has an euler path */
+igraph_integer_t is_eulerian_undirected(igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
```
Functions that return an error code must always have return type `int`.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        IGRAPH_CHECK(igraph_es_incident(&incoming, i, IGRAPH_IN));
```
If all you need is the degree, simply use `igraph_degree`

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+    }
+
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_strong, IGRAPH_STRONG));
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_weak, IGRAPH_WEAK));
```
Why do you check both weak and strong connectivity? Strong connectivity is required, weak connectivity is irrelevant.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_strong, IGRAPH_STRONG));
+    IGRAPH_CHECK(igraph_is_connected(graph, &res_weak, IGRAPH_WEAK));
+
+    if ((outgoing_excess == 0 && incoming_excess == 0) && (res_strong == 1)) {
```
Do not use `==` with Boolean variables, as it hurt readability. Furthermore, `if (x == 1)` is NOT equivalent to `if (x)`, as any non-zero value encodes "true". Please change these (and all the other things I brought up in individual comments) throughout the code, not just in the location where I added the comment.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,502 @@
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    } else if ((outgoing_excess == 1 && incoming_excess == 1) && 
+        (res_strong == 1 || res_weak == 1)) {
```
This is incorrect. Weak connectivity is insufficient. Also, `res_strong || res_weak` is the same as `res_weak`.

iosonofabio replied:
This changed quite a bit since. Plus, you don't need strong connectivity for a Eulerian path (though you do for a Eulerian cycle): the example is the graph 0->1 which is not strongly connected (since there's no reaching 0 from 1) but has a Eulerian path (namely, the only edge).

I clarified this with some comments now

YasirKusay commented:
Thank you for all your feedback. I will continue refining my implementation tomorrow.

YasirKusay commented:
Hello, just a quick update I have decided that rather than deleting edges during the eulerian walk, I will instead have a list of visited and unvisited edges. A visited edge will essentially be similar to a "deleted" edge and the advantages of this are that I will avoid issues to do with the edge id's being changed when the edges are deleted.

YasirKusay commented:
Hello, I have just finished implementing the changes that I said I would do in my last comment on this pull request. It appears to be working fine but I will need to do further tests to ensure that this is working correctly.

szhorvat commented:
Let us know when all the comments I made previously have been addressed. I'll take a more detailed look then.

YasirKusay commented:
I believe that I have addressed all your comments. However, with respect to the comment you made about weak connectivity in a directed graph being insufficient is not necessarily true. For example, 0->1->2  is not a strongly connected graph, but it does have an eulerian path. While it is true that for a directed graph to have a circuit, it must be strongly connected, a directed graph that has a path does not need to be strongly connected (0->1->2).

szhorvat commented:
Yes, you are correct. Sorry about that. You can respond to comments directly. Please do respond directly, even if you believe that a comment was addressed, so we can more easily keep track.

However, I see several unaddressed comments. Can you please address all of them, including seemingly cosmetic ones? Do not introduce unnecessary variables in tests. Remove all unneeded `#include`s, there is still a comment about `is_eulerian_undirected` returning 0, 1 or 2, etc. It would be great if you could try to make the PR merge-ready.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+
+/* solution adapted from https://www.geeksforgeeks.org/eulerian-path-and-circuit/
+The function returns one of the following values
+0 -> if graph is not Eulerian
```
Please clean up outdated comments.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+    *has_path = 0;
+    *has_cycle = 0;
+
+    IGRAPH_FINALLY_CLEAN(2);
```
There are a few memory management issues in this function. The `IGRAPH_FINALLY_CLEAN` should always be right after the actual cleanup. Here the function can potentially return before running the `IGRAPH_FINALLY_CLEAN`, or even before running the `igraph_vector_destroy` lines (!!). It is important to be very careful about this. Please double-check all `return`s and make sure that there cannot be any leaks.

szhorvat replied:
Sorry, I see that you have duplicated the `igraph_vector_destroy` on line 134, so the memory does get freed. However, the `IGRAPH_FINALLY_CLEAN` gets skipped, which is a problem.

Perhaps you could move the cleanup to the end of the function, and use a `goto` to jump there when needed. This would make the function cleaner, as all cleanup code would be in one place.

YasirKusay replied:
I changed my code such that igraph_finally_clean gets executed before the first return in that function.

iosonofabio replied:
These things got checked and we also have a FINALLY check at the end of the tests now.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+    return IGRAPH_SUCCESS;
+}
+
+/* flexible function */
```
What does this comment refer to? Remember, very likely people who don't know you will read and modify this code in the future. They won't be able to ask you. The comments are addressed to them.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+    return 1;
+}
+
+int print_euler_undirected_implementation(igraph_integer_t start, igraph_t *g, igraph_vector_t *path, igraph_vector_bool_t *visited_list) {
```
Why does this function have print in the name?

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+    return IGRAPH_SUCCESS;
+}
+
+int igraph_eulerian_path(igraph_t *graph, igraph_vector_t *res) {
```
Can you please draft documentation for all the public functions?

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+
+    igraph_is_eulerian(graph, &has_path, &cycle);
+
+    if (!has_path && !cycle) {
```
Why is it necessary to check `&& ! cycle`?

YasirKusay replied:
Good comment. If a eulerian cycle exists, then that also means a path exists. I have decided to have that function (eulerian_path) return an eulerian cycle (if it exists).

YasirKusay replied:
When I was writing this, I had my is_eulerian function have has_path equal to zero if a cycle exists. That is why I had that check. I feel however that my implementation may not make sense would you agreee?

iosonofabio replied:
Let's change it such that when cycle is 1, has_path is also 1.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+    igraph_is_eulerian(graph, &has_path, &cycle);
+
+    if (!cycle) {
+        IGRAPH_WARNING("Eulerian cycle not possible");
+        return IGRAPH_FAILURE;
```
Should be an error, not a warning. Use EINVAL as the error code, as it is more specific than FAILURE. You can check the available error codes in https://github.com/igraph/igraph/blob/master/include/igraph_error.h

YasirKusay replied:
I did take a look at the potential errors but I was not sure at what the appropriate error code would be in this case. Could you recommend me one?

szhorvat replied:
> Could you recommend me one?

I did: EINVAL.

szhorvat replied:
I think you misunderstood. Just use `IGRAPH_ERROR` and no `return`. The `IGRAPH_ERROR` macro will do the return (see docs, https://igraph.org/c/doc/igraph-Error.html)

P.S. Originally mentioned here:

https://github.com/igraph/igraph/pull/1346#discussion_r397742443

iosonofabio replied:
This is fixed now

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,499 @@
+
+    igraph_is_eulerian(graph, &has_path, &cycle);
+
+    if (!has_path && !cycle) {
+        IGRAPH_WARNING("Euler cycle not possible");
+        return IGRAPH_FAILURE;
+    }
```
Is this necessary? It seems to me that all public functions that call this already do the check. Did I miss anything?

YasirKusay replied:
My bad, forgot to remove this.

szhorvat commented:
@iosonofabio Could you please look through this PR when @YasirKusay says that it is ready?

@YasirKusay It might be a good idea to look through all the old individual comments I made, to see if anything was missed (I would need to do that too, as I do not remember all of them). Just check above in this thread, and make sure you click the "Load more" button to see all of them.

YasirKusay commented:
@szhorvat 
I will do that. Thank you for your reviews.

szhorvat commented:
I have some pressing deadlines at the moment, but I'll take a detailed look at the algorithm during the coming couple of weeks. Thanks for all the work you put into this @YasirKusay !

iosonofabio commented:
@szhorvat Yasir and I had a chat just yesterday.

Honestly this PR raises a warning for igraph development I think. In this particular case, I am not sure going back through 80 messages looking for unaddressed comments is particularly robust, but we'll do it anyway because that's the way things are now - your time is greatly appreciated as well @szhorvat, thank you for that.

However it is very difficult for new developers, despite a good training in CS, to contribute to igraph if every PR needs to be pitch perfect before merging. I would advocate for lowering that bar and going through a few routine checks as a second step before merging the development branch into master. Anyway, I'll open a thread on the discourse.

iosonofabio commented:
Here a list of things that still need fixing:

- [ ]

szhorvat commented:
>  pass tests on AppVeyor

Just a note that something is wrong with Appveyor, [or rather msys2](https://github.com/msys2/MSYS2-packages/issues/1962), at this moment. Hopefully it'll fix itself in a few days.

ntamas commented:
Regarding Appveyor: you need to install `zstd` in the build setup; some of the packages that MSYS2 tries to install are compressed with `zst`, but it is not installed by default. I believe I fixed it already in the master branch, but this PR was forked off before that.

szhorvat commented:
To avoid confusion: the response from @ntamas was to a post I made earlier today, but then deleted (as I didn't want a link back to here from the msys2 repo, where I referenced a bug). 

I think Appveyor will be broken for a few days anyway, see https://igraph.discourse.group/t/appveyor-currently-broken/245/5

szhorvat commented:
@iosonofabio @YasirKusay One more for the TODO list: could you please [change the target to the `develop` branch](https://stackoverflow.com/a/24159161) and pull in the latest changes from there? This will also pull in the latest test configuration.

YasirKusay commented:
Hi, I just saw your messages and will do these things after I come from work tomorrow.

szhorvat commented:
Just a small comment:

```
 \example examples/test/igraph_is_eulerian.c
```

Please do not add any tests to the documentation. We are trying to separate examples from tests. It's okay not to have a usage example in the docs.

iosonofabio commented:
@szhorvat we could use a little advice on the bridge checks. We used to compute all bridges for every edge, which is inefficient. We obviously don't need to check for bridge/no bridge if this is the only unvisited edge out of a vertex because we're going that way anyways.

However, if there are several unvisited edges we need to do a DFS from the current vertex with/without the current edge to figure out whether **this** particular edge is now a bridge. Notice the edge might not have been a bridge initially, but could have become a bridge once a few other edges have been visited.

I see that igraph has `igraph_dfs`, however it's not used much internally, instead the recursion is written down explicitely (e.g. in the bridges function itself). How would you go about this? Shall we call `igraph_dfs` for readability or code out the recursive DFS algorithm again?

Thanks!

szhorvat commented:
@iosonofabio I must beg for some patience until next week because of looming deadlines ... I must go through the code in detail to be able to answer this well.

A question:

The comments in the current version suggest that you wanted to implement Fleury's algorithm for the undirected case and Hierholzer's algorithm for the directed case. Is this the case? If yes, why not use Hierholzer's for the undirected case as well? It is much more efficient, and it avoids all the trouble with the bridges. It is possible that I am missing an important point here.

Regarding Fleury's algorithm:

The reason why it is a good idea to avoid both `igraph_bridges` and `igraph_dfs` is that these functions take an `igraph_t` as input. Then internally they convert it to a different representation (some sort of adjacency list) and operate on that.  Doing this conversion repeatedly seems like a waste. Furthermore, since these functions take an `igraph_t` data structure as input, we would be stuck with having to delete edges explicitly from the `igraph_t` with functions like `igraph_delete_edges`. This is also very expensive (linear time for a single deletion).

Usually, the better approach is to decide what representation is the best to use for a given algorithm, convert the graph to that representation *once*, then use it throughout the function. This is what most igraph functions do currently. This is more work, as graph traversals like DFS must be implemented separately for each function / each representation. But it gives us the option to make significant optimizations. For example, an edge can be simply *marked as deleted/handled* (O(1) time) instead of actually deleting it from the `igraph_t` data structure (which takes O(n) time). Then we can write our specialized DFS to simply avoid edges marked as deleted. Or, we can use a data structure in which edge deletion is cheaper (an incidence list with an appropriate storage container for edges).

What I said above is just the principle of the matter. I tried to explain why you almost never see `igraph_dfs` used explicitly.

To give a proper answer for this specific case, I must look through the code in detail first. This is why I asked for more time.

iosonofabio commented:
Awesome, thanks for the info. I wonder whether there might be a way to have a generic but efficient DFS function which for one or two typical representations. Anyway not for this PR.

@YasirKusay let's look into Hierholzer for undirected as well first.

@szhorvat let us try and use that algorithm for everything first. No need to spend more time on this until we get back to you. Thank you!

szhorvat commented:
> I wonder whether there might be a way to have one or two generic but efficient DFS functions for typical representations of the graph.

Yes, personally I am a bit frustrated by this situation. The internal data structure is very commonly an igraph_adjlist or an igraph_inclist. Sometimes, one might need some functionality that is already implemented, and it would be very convenient to be able to run this *directly* on that adjlist or inclist. Actually, a few internal functions seem to take an adjlist (not an igraph_t) as input. Sometimes I wish that more of them had this.

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,612 @@
+
+    for (i = 0; i < e_count; i++) {
+        VECTOR(visited_list)[i] = 0;
+    }
```
Default initialization might have zero everywhere already? just a little note

ntamas replied:
Even if it is not, we have `igraph_vector_fill()`. But I think the default is to have zeros everywhere as we use `calloc()` for memory allocation.

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,612 @@
+        if (VECTOR(*outgoing_list)[curr] != 0) {
+            IGRAPH_CHECK(igraph_stack_push(&tracker, curr));
+            
+            IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
```
probably put this out of the while?

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,612 @@
+            
+            igraph_inclist_destroy(&il);
+            IGRAPH_FINALLY_CLEAN(1);
+
```
see above, probably no need to create and destroy `il` every time?

ntamas commented:
> Yes, personally I am a bit frustrated by this situation. The internal data structure is very commonly an igraph_adjlist or an igraph_inclist.

True, true. In theory, we could replace the current graph representation in `type_indexededgelist.c` with an adjacency list or an incidence list (or both, in parallel). What I don't know yet is how many other algorithms are there in the library that benefit more from the current indexed edge list representation and that would become slower if we switched to an adjacency list. The best would be to create a comprehensive benchmarking suite for igraph that reflects typical use-cases, and then we could measure what happens if we replace the current representation with another one.

iosonofabio commented:
@szhorvat Yasir has rewritten the undirected graph algorithm to use Hierholzer and it now works fine. Tests on Linux/OSX pass, tests on AppVeyor seem to pass, although a few are still ongoing. Yasir will ping us tonight if everything is fine and then it'd be great to have you or @ntamas review the code.

We are so close!!

Thanks,
Fabio

iosonofabio commented:
@YasirKusay Appveyor passes except for Python3.8 because it cannot find the "conda" command. That seems unrelated to this PR, so I'd say the PR is ready for review!

ntamas commented:
Yes, that's unrelated; I tried to add Python 3.8 to the test environment on Appveyor but it would need a newer Appveyor image (the Visual Studio 2013 image does not have Python 3.8 yet).

However, when you think about it: why do we even test the C build for various versions of _Python_? Could be an artifact from the times when we did not have proper CI setup in `python-igraph`; I need to revise whether it's necessary at all.

szhorvat commented:
We still need to add this to the docs. @iosonofabio, any suggestions on where it would fit? We can add ti to the Structural Properties chapter, but that chapter is already overloaded and incoherent. I would not give this topic its own chapter though, it's too short. 

Perhaps start a chapter on cycles? Later we can move [tha DAG section](https://igraph.org/c/doc/igraph-Structural.html#idm383442187344) there, and eventually also add the cycle basis there. If/when we get any Hamiltonian cycle implementation, that will fit too.

szhorvat commented:
I'll take a careful look during this week and will post a review.

@YasirKusay I'm happy to integrate the docs once we come to a decision, you don't need to spend time on that.

iosonofabio commented:
A new one on cycles sounds great. Structural props really needs a little love.

Thank you so much!

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+    has_path = 0;
+    cycle = 0;
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+    
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    IGRAPH_CHECK(igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS));
+
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    for (i = 0; i < igraph_vector_size(&degree); i++) {
+        if (((long int) VECTOR(degree)[i]) > 0) {
+            start = i;
+            break;
+        }
+    }
+
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(1);
```
Computing degrees twice (`degree` and `out_degree`) appears to be unnecessary.

The naming `out_degree` is confusing given that this function deals with undirected graphs only.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+
+    igraph_inclist_destroy(&incl);
+    igraph_vector_destroy(&out_degree);
+    IGRAPH_FINALLY_CLEAN(1);
```
This should be 2, not 1.

Please be extremely careful with this because mistakes will lead to hard-to-debug crash. My test program was crashing and it took me quite a while before I managed to find out that it was due to this line. Count the number of `IGRAPH_FINALLY`s and make sure that the match the the `IGRAPH_FINALLY_CLEAN`

ntamas replied:
One trick that we could adopt is that we could check whether the "finally" stack is clean at the end of the unit tests. If it is not, chances are that `IGRAPH_FINALLY_CLEAN()` was invoked with a wrong count somewhere.

See `examples/simple/igraph_strvector.c` for an example; it has something like this at the end:

```
if (!IGRAPH_FINALLY_STACK_EMPTY) {
    /* run in circles, scream and shout */
}
```

iosonofabio replied:
Excellent catch @szhorvat and idea @ntamas, I'm in favor

szhorvat replied:
We could have a macro in the test utilities to help with this and make it a one-liner at the end of every test program.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+}
+
+
+int igraph_i_eulerian_path_undirected_implementation(igraph_integer_t start_node, igraph_t *graph, igraph_vector_t *res, igraph_vector_t *outgoing_list) {
```
Let's always make the "this" argument the first one. That's the `igraph_t *graph` in this case.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &incl, IGRAPH_ALL));
```
Why do we need to build an `inclist` here?

It seems to me that this is just a 3rd way to compute the degrees, and is unnecessary.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+                break;
+            }
+        }
+        igraph_i_eulerian_path_undirected_implementation(start, graph, path, &out_degree);
```
IGRAPH_CHECK is missing.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+    if (cycle && !has_path) {
+        IGRAPH_CHECK(igraph_i_eulerian_path_undirected_implementation(start, graph, path, &out_degree));
+    } else {
+        for (i = 0; i < igraph_vcount(graph); i++) {
+            incedges = igraph_inclist_get(&incl, i);
+            if ( igraph_vector_int_size(incedges) % 2 == 1) {
+                start = i;
+                break;
+            }
+        }
```
Let's move the `start` calculation to one place, together with line 321 above. Then we can move the call to `igraph_i_eulerian_path_undirected_implementation` out of the `if`/`else` block, and avoid duplicating it.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+                }
+            }
+            
+            next = IGRAPH_TO(graph, edge) == curr ? IGRAPH_FROM(graph, edge) : IGRAPH_TO(graph, edge);
```
You can use `next = IGRAPH_OTHER(graph, edge, curr)`

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+        if (VECTOR(*outgoing_list)[curr] != 0) {
+            IGRAPH_CHECK(igraph_stack_push(&tracker, curr));
+            
+            incedges = igraph_inclist_get(&il, curr); /***** INSPECT LATER */
```
What does this comment refer to?

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+    }
+
+    while (!igraph_stack_empty(&edge_path)) {
+        IGRAPH_CHECK(igraph_vector_push_back(res, igraph_stack_pop(&edge_path)));           
```
You are pushing to the vector `res`, but it has not been cleared. What if the user passes a non-empty vector?

`res` should be [cleared](https://igraph.org/c/doc/igraph-Data-structures.html#igraph_vector_clear) at the start of the function.

We know exactly how many elements the result will have. Therefore, let us [reserve](https://igraph.org/c/doc/igraph-Data-structures.html#igraph_vector_reserve) precisely that amount of storage at the same point where the vector is cleared.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,614 @@
+    }
+
+    while (!igraph_stack_empty(&edge_path)) {
+        IGRAPH_CHECK(igraph_vector_push_back(res, igraph_stack_pop(&edge_path)));           
```
Please see comment on the directed implementation of the same.

YasirKusay commented:
Hello, I have addressed all of the things that you have requested. As for the very basic cases (graphs with 0 vertices, etc), thank you for pointing it out! I agree with you with most of the things, except one case, that being a graph with 1 vertex. If a graph has 1 vertex and 0 edges, it should have an empty return, whenever igraph_eulerian_cycle or igraph_eulerian_path is called, but not when there are edges. E.g. a graph with one vertex (vertex 0) that has one or more self loops should not have an empty cycle/path, rather should have a normal cycle/path return.

szhorvat commented:
> If a graph has 1 vertex and 0 edges, it should have an empty return, whenever igraph_eulerian_cycle or igraph_eulerian_path is called, but not when there are edges. E.g. a graph with one vertex (vertex 0) that has one or more self loops should not have an empty cycle/path, rather should have a normal cycle/path return.

You are of course perfectly right!

I notice that in the `is_eulerian` functions now you use `(igraph_vcount(graph) == 1 && igraph_ecount(graph) == 0)`. However, if there is a single vertex only, both `has_cycle`and `has_edge` are always true. Thus, this check:


    if (igraph_vcount(graph) == 0 || (igraph_vcount(graph) == 1 && igraph_ecount(graph) == 0) || (igraph_vcount(graph) > 0 && igraph_ecount(graph) == 0)) {
        *has_path = 1;
        *has_cycle = 1;
        return  IGRAPH_SUCCESS;
    }


could be simplified to 


    if (igraph_ecount(g) == 0 || igraph_vcount(g) <= 1) {
        *has_path = 1;
        *has_cycle = 1;
        return  IGRAPH_SUCCESS;
    }


Do you agree @YasirKusay ?

szhorvat commented:
In `igraph_i_is_eulerian_directed`, you check both strong and weak connectivity. I think it is sufficient to check weak connectivity (more precisely, that there is precisely one weakly connected component that has size > 1.

Weak connectivity plus the fact that all vertices have equal in- and out-degree (except maybe the starting and the ending vertex) implies strong connectivity.

Do you agree @YasirKusay? Let me know if I missed something.

szhorvat commented:
Another small improvement that could be made is to have `igraph_i_is_eulerian_directed` and `igraph_i_is_eulerian_undirected` (the *internal* versions) compute and return the start and end vertices (for the case when we have a path). They *have* to do this anyway when checking that there is only one vertex with `outdegee - indegree == 1` and with `indegree - outdegree == 1`. 

Thus you can save re-calculating the start/end vertices a second time in the function that finds the path/cycle.

szhorvat commented:
`igraph_i_eulerian_path_directed` appears to duplicate the check of whether the graph is Eulerian. You already do this check *before* calling `igraph_i_eulerian_path_directed`. Do not do it a second time.

YasirKusay commented:
Hello, thanks for the extensive feedback and suggestions! As for your query regarding checking strongly connected clusters, I did it because I need to consider the disconnected graph case. However, thanks for linking that stackexchange post, as indegree == outdegree for all nodes does nullify the need to check if a graph is strongly connected (but as you pointed out, the weakly connected check is still necessary to ensure that there are no 2 disconnected clusters, each containing more than 1 edges). 
You are also very correct in pointing out that checking whether a graph is strongly/weakly connected does not need to be carried out in igraph_i_eulerian_directed, as by this point, we are now certain that the graph contains either a graph or cycle. I have modified it as such and removed the need to check clusters and whether the graph is strongly and weakly connected.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+has_path is set to 1 if a path exists, 0 otherwise
+has_cycle is set to 1 if a cycle exists, 0 otherwise
+*/
+int igraph_i_is_eulerian_undirected(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
```
Please make every function that is only used within this translation unit `static`.

iosonofabio replied:
done by now

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+    igraph_integer_t curr = start;
+    igraph_integer_t next, e_count, curr_e;
+    igraph_inclist_t il;
+    igraph_stack_t path, tracker, edge_tracker, edge_path;
```
As far as I can tell, `path` is not used by the algorithm and not returned. @YasirKusay can you please go through the entire source file and remove any leftover code or variables which are no longer used?

szhorvat replied:
Scratch that, see my comment in the main thread below.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+
+    IGRAPH_CHECK(igraph_clusters(graph, NULL, &csize_weak, &clusters_weak, IGRAPH_WEAK));
+    cluster_count = 0;
+    for (i = 0; i < igraph_vector_size(&csize_weak); i++) {
+        if (VECTOR(csize_weak)[i] > 1) cluster_count++;
+    }
```
We also need to remember the special case of a graph with more than one vertex, but having self-loops on only one of them.

This is Eulerian:

![image](https://user-images.githubusercontent.com/1212871/84040672-2f817c80-a9a3-11ea-8feb-73b24d99d824.png)


This is not:

![image](https://user-images.githubusercontent.com/1212871/84040631-22fd2400-a9a3-11ea-87b1-974ed51aff8f.png)

This is also Eulerian:

![image](https://user-images.githubusercontent.com/1212871/84040708-3ad4a800-a9a3-11ea-837b-ee929edb1bbd.png)

Please add these to the test cases, as these are easy to get wrong, and someone might break them in the future

iosonofabio replied:
we check for those cases now

szhorvat commented:
@YasirKusay Earlier you proposed to return both the edge-path and the vertex-path. I said no to that, as I thought it was unnecessary.

Thinking about it more, it may make sense to return that as well, or at least optionally return the starting vertex of the path. If the graph is undirected, it is obvious which one is the starting vertex. But if it is not, it takes extra work to find it (the odd-degree endpoint of the starting edge).

Consider a graph like this,

![image](https://user-images.githubusercontent.com/1212871/84045558-3ad7a680-a9a9-11ea-95be-aadda5994270.png)

The vertex path is 1-2-1-2-3.

The edge path could be 1-2-3-4. It's a bit hard to see how that actually goes.

@iosonofabio @vtraag Any opinions on this?  

Since @YasirKusay's code already computes the vertex path (I added a comment above asking to remove it), maybe let's go with Yasir's original suggestion, and make it possible to get both the vertex and edge paths? If the user doesn't want one of them, they should be able to pass `NULL`.

iosonofabio commented:
@szhorvat there's a reason I suppose why the whole Eulerian business is defined on edges and your lot makes for a good case. On the other hand, with multiedges it doesn't matter which edge you choose, both choices lead to a Eulerian path/cycle. So I'm fine either way, as long as it's deterministic.

Also, @szhorvat since you prefer to fix the last few bolts and nuts yourself, feel free to push to the PR from now on!

vtraag commented:
> make it possible to get both the vertex and edge paths?

Sounds reasonable to me! Indeed, as you say, a user can always pass `NULL`, and I imagine that it does not degrade performance.

szhorvat commented:
I merged the recently added tools that verify the FINALLY stack size. It appears that there are some more mismatches between the number of `IGRAPH_FINALLY` and `IGRAPH_FINALLY_CLEAN`.

YasirKusay commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+        if (VECTOR(in_degree)[i] != VECTOR(out_degree)[i]) {
+            if ((VECTOR(in_degree)[i] + 1 == VECTOR(out_degree)[i]) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+            } else if ((VECTOR(out_degree)[i] + 1 == VECTOR(in_degree)[i]) && (incoming_excess < 1 && outgoing_excess < 2)) {
```
The reason why I did that is because outgoing_excess should be either 0 or 1, but we do not know which one it is, thats why I did that. The same applies to incoming_excess.

szhorvat replied:
I do not understand what you mean. As you say, `outgoing_excess` is either 0 or 1. The condition `outgoing_excess < 2` is always true. Why do you include the condition?

iosonofabio replied:
fixed now

YasirKusay commented:
Just a quick update, I am still working on this but will inform you of when I am done. Although I fixed the issue to do with a graph having edges: 0,0, 1,1, the fix is inefficient and I will think about a potentially faster method when I am able to work on this repo next time.

szhorvat commented:
@YasirKusay Haven't looked at your new changes yet, just wanted to ask: Will it help if I push an efficient function that checks the connectivity of the line graph (i.e. the type of connectivity that is relevant here)?

YasirKusay commented:
I wouldn't mind that @szhorvat. Any help would be appreciated.

szhorvat commented:
I'll do that, give me a couple of days.

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+            IGRAPH_CHECK(igraph_stack_push(&tracker, curr));
+            
+            incedges = igraph_inclist_get(&il, curr);
+            nc = igraph_vector_int_size(incedges);
```
I gave this a pass

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+
+int igraph_i_is_eulerian_directed(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_bool_t res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess, clusters_weak;
```
gone

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+int igraph_i_is_eulerian_directed(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_bool_t res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess, clusters_weak;
+    int i, cluster_count;
```
changed into long int

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_stack_init(&path, igraph_vcount(graph)));
```
this is done by now

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        if (VECTOR(in_degree)[i] != VECTOR(out_degree)[i]) {
+            if ((VECTOR(in_degree)[i] + 1 == VECTOR(out_degree)[i]) && (incoming_excess < 2 && outgoing_excess < 1)) {
```
I fixed this whole section

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+            start = i;
+            break;
+        }
+    }
```
seems like this was done

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    /* determining the start node */
```
seems like was done already

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+    cluster_count = 0;
+    for (i = 0; i < igraph_vector_size(&csize); i++) {
+        if (VECTOR(csize)[i] > 1) cluster_count++;
+    }
```
vector size moved out of the loop now

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+    res_weak = 1;
+
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
```
fixed this

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+        }
+        IGRAPH_CHECK(igraph_i_eulerian_path_directed_implementation(graph, &start_node, &out_degree, res));
+        igraph_vector_destroy(&degree);
+        IGRAPH_FINALLY_CLEAN(1);
```
I think this got fixed meanwhile

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+
+
+/* solution adapted from https://www.geeksforgeeks.org/hierholzers-algorithm-directed-graph/ */
+int igraph_i_eulerian_path_directed_implementation(const igraph_t *graph, igraph_integer_t *start_node, igraph_vector_t *outgoing_list, igraph_vector_t *res) {
```
this function got merged with the other one, so it got solved too

iosonofabio commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,565 @@
+
+
+int igraph_i_eulerian_path_undirected_implementation(const igraph_t *graph, igraph_integer_t start_node, igraph_vector_t *res, igraph_vector_t *outgoing_list) {
+    igraph_integer_t start = start_node;
```
this variable got renamed

iosonofabio commented:
Hi guys, I did a buncha changes that address each and every comment in the last half a year.

Although this PR is not and never will be a perfect exercise of style, I think this is ready for merging.

Please let me know if you find issues that are not caught by the relatively extensive tests and I'll be happy to fix them. Otherwise, let's merge it ;-)

ntamas commented:
Thanks for your hard work for all of you guys (@YasirKusay @iosonofabio @szhorvat )! I'll wait for a final nod from @szhorvat as he was doing most of the reviewing - if he is okay with things, I'll be happy to merge this.

iosonofabio commented:
Perfect! As I said, happy to implement any reasonable changes :-)

szhorvat commented:
@iosonofabio It's really busy times right now for me, so I'll post here what I notice gradually. I'd like to try out the functionality again and run a few tests before merging.

Two things I noticed:

 - The docs need to be updated to describe the new argument
 - Instead of `vector_res`, it's clearer to use `vertex_res`, as this vector contains vertex IDs (right?)

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,652 @@
+
+    if (vector_res) {
+        igraph_vector_clear(vector_res);
+        IGRAPH_CHECK(igraph_vector_reserve(vector_res, n+1)); 
```
I believe this should be `m+1`, not `n+1`. The final size of this vector will be `m+1`.

ntamas replied:
Addressed in 2d36c3c

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,652 @@
+
+    if (vector_res) {
+        igraph_vector_clear(vector_res);
+        IGRAPH_CHECK(igraph_vector_reserve(vector_res, n+1));
```
Same as above, this should be `m+1`, not `n+1`.

ntamas replied:
Addressed in 2d36c3c

ntamas commented:
Docs updated and `vector_res` renamed to `vertex_res` in 2d36c3c (I had a bit of spare time)

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+        }
+    }
+
+    /* assuming that cluster count is not greater than 1 now */
```
Is this comment meant to come before the previous loop?

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+        }
+    }
+
+    /* assuming that cluster count is not greater than 1 now */
+    /* however, we need to check for self loops still*/
```
Were these comments meant to come before the previous loop?

iosonofabio replied:
moved that

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    vsize = igraph_vector_size(&check_for_self_loops);
+    for (i = 0; i < vsize; i++) {
+        igraph_vector_int_t *incedges = igraph_inclist_get(&il, VECTOR(check_for_self_loops)[i]);
+        long int nc = igraph_vector_int_size(incedges);
+        if (nc > 0) {
+            *has_path = 0;
+            *has_cycle = 0;
+            igraph_vector_destroy(&csize);
+            igraph_inclist_destroy(&il);
+            igraph_vector_destroy(&check_for_self_loops);
+            igraph_vector_destroy(&cluster_member);
+            IGRAPH_FINALLY_CLEAN(4);
+
+            return IGRAPH_SUCCESS;
+        }
+    }
```
Why do we need to construct an inclist here? We already compute the degree vector below. As far as I can tell, we could simply use the degree vector to check for self-loops.

Also: this was with `IGRAPH_OUT`, but the graph is undirected. It's not wrong, but a bit confusing (for a cminute I thought I was looking at the directed version).

iosonofabio replied:
I refactored this whole section using only degrees

szhorvat commented:
The result for graphs that contain one vertex with one or more self-loops and several other zero-degree vertices is incorrect. The functions say that this graph is not Eulerian, but it is.  Example graph:

<img width="96" alt="image" src="https://user-images.githubusercontent.com/1212871/86514235-3b7d2480-be11-11ea-9c31-31db283345ea.png">

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+        *has_cycle = 0;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    /* check if singleton self loops destroy the Eulerian path/cycle */
+    vsize = igraph_vector_size(&check_for_self_loops);
+    for (i = 0; i < vsize; i++) {
+        igraph_vector_int_t *incedges;
+        incedges = igraph_inclist_get(&il, VECTOR(check_for_self_loops)[i]);
+        long int nc = igraph_vector_int_size(incedges);
+        if (nc > 0) {
+            *has_path = 0;
+            *has_cycle = 0;
+        }
+    }
```
This is basically checking the degrees. Do not construct an inclist to do this. The degrees have already been computed with `igraph_degee` earlier.

iosonofabio replied:
I redesigned it using only degrees, good point

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+    igraph_vector_t out_degree, in_degree, csize_weak, check_for_self_loops, cluster_member;
+    igraph_inclist_t il;
+
+    if (!graph) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    }
```
Can you explain why this is needed? The top level function does not (should not) allow passing NULL for the graph.

iosonofabio replied:
Indeed, internal functions never check for this, so I removed it

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+
+    n = igraph_vcount(graph);
+
+    if (igraph_ecount(graph) == 0 || n <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    }
```
Whenever the function returns with `has_path` being true, it should not leave `start_of_path` un-set.

szhorvat replied:
Just fixed this.

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,660 @@
+    long int i, j, vsize;
+    long int cluster_count;
+
+    if (igraph_ecount(graph) == 0 || igraph_vcount(graph) <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return  IGRAPH_SUCCESS;
+    }
```
Whenever the function returns with `has_path` being true, and the graph has at least one vertex, it should not leave `start_of_path` un-set. The graph may have one vertex with many self-loops, so `start_of_path` may end up being used.

szhorvat replied:
Just fixed this.

iosonofabio commented:
Alright guys, here's the next iteration.

@szhorvat pointed out a bunch of problems that honestly I should have spotted myself. I missed the forest for the trees while polishing the code.

Now I redesigned both undirected and directed functions for `is_eulerian` using a parallel and simplified logic, a new test covering the corner case that was raised, and sometimes better variable names. I also reorganized the order of tests.

I think this should be good by now - all tests pass. If @szhorvat and @ntamas could both have a look one of the next days then we can close :-)

thank you!

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,671 @@
+    IGRAPH_FINALLY_CLEAN(1);
+
+    /* the graph is connected except for singletons */
+    /* find singletons (except for self-loops) */
```
"(except for self-loops)" <- Here you mean to say, "including those with self-loops", right? Or am I misunderstanding?

szhorvat commented on src/igraph_eulerian.c:
```diff
@@ -0,0 +1,671 @@
+    long int cluster_count;
+    /* number of self-looping singletons */
+    long int es;
+    /* number of non-singletons with edges */
```
There can't be a non-singleton vertex _without_ edges. Also, this variable isn't actually used to count such vertices. It is simply set to 1 if there is at least one such vertex.

Overall, the algorithm seems correct, but I was a bit confused by the comments.

szhorvat commented:
Merged 🥳, thanks @YasirKusay and @iosonofabio !

szhorvat commented:
@iosonofabio I clarified some comments before merging (no code changes). If you disagree, please just fix it directly in develop.

szhorvat commented:
@iosonofabio There's some weirdness in how doc comments work, so paragraphs must be separated using `</para><para>`, _without_ an opening `<para>`.

Fixed it here: https://github.com/igraph/igraph/commit/313bdc7991a1ecff5cc2402a5a5dda6b2cfa9507

iosonofabio commented:
Thank you, happy to see it done! Yes a few comments were a little off, including the paragraph thing. No problem from my side!

On Tue, Jul 14, 2020, at 04:00, Szabolcs Horvát wrote:
> 
> 
> @iosonofabio <https://github.com/iosonofabio> There's some weirdness in 
> how doc comments work, so paragraphs must be separated using 
> `</para><para>`, *without* an opening `<para>`.
> 
> Fixed it here: 313bdc7 
> <https://github.com/igraph/igraph/commit/313bdc7991a1ecff5cc2402a5a5dda6b2cfa9507>
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/igraph/igraph/pull/1346#issuecomment-657705912>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJFEAAEECMNI23GP5JQGVDR3NDSZANCNFSM4K7L7OIA>.
>

YasirKusay commented:
I am happy to see that this pull request is now closed, being a relatively new programmer, this definitely provided me with a lot of experience and allowed me to improve as a programmer. Thank you all for the opportunity, and for helping me out with this pull request.
